### PR TITLE
Documentation/sql getalldestinations comments

### PR DESCRIPTION
**Pull request by passionfruit18**
Made AddToSet symmetrical, added documentation comments explaining GetAllDestinations, and formatted long single line queries into multiple lines to make reading easier.

### CODE REVIEW

File: ModelCatalogueCorePluginTestApp/src/groovy/org/modelcatalogue/core/mysql-procedures.sql [hunk 1]
```diff
@@ -7,30 +7,48 @@ DETERMINISTIC
...
             join catalogue_element dest on dest.id = r.destination_id
-            where r.relationship_type_id = LeafType or r.relationship_type_id = DescendType and src.data_model_id <> dest.data_model_id);
-        SET rv = '';
+            where r.relationship_type_id = LeafType or
+                  (r.relationship_type_id = DescendType and src.data_model_id <> dest.data_model_id));
+
+        SET rv = ''; # This is the result set
         SET queue = InitialQueue;
         SET queue_length = LENGTH(InitialQueue) - LENGTH(REPLACE(InitialQueue, ',', '')) + 1;
+        # a clever trick for finding the length of a comma separated list
```
Comment (musketyr) [inline ModelCatalogueCorePluginTestApp/src/groovy/org/modelcatalogue/core/mysql-procedures.sql]: I think to put the comments to the next line after the statement makes the association difficult. I would prefer format

```
# this is a comment
this is a line commented
```

**Comment by musketyr** [comment]
there is something the MySQL doesn't like:

```
[1m> Building 90% > :runGrailsTests[22m[32D2017-04-13 12:37:32,423 [localhost-startStop-1] ERROR context.GrailsContextLoaderListener  - Error initializing the application: Failed to execute SQL script statement #3 of resource class path resource [org/modelcatalogue/core/mysql-procedures.sql]: CREATE FUNCTION AddToSet(NextValues MEDIUMTEXT, CurrentSet MEDIUMTEXT) RETURNS MEDIUMTEXT DETERMINISTIC BEGIN IF LENGTH(CurrentSet) = 0 OR CurrentSet is NULL THEN RETURN NextValues; ELSEIF LENGTH(NextValues) = 0 OR NextValues is NULL THEN RETURN CurrentSet; # AddToSet should be symmetrical ELSE RETURN CONCAT(CurrentSet,',',NextValues); END IF; END ; nested exception is com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 1
org.springframework.jdbc.datasource.init.ScriptStatementFailedException: Failed to execute SQL script statement #3 of resource class path resource [org/modelcatalogue/core/mysql-procedures.sql]: CREATE FUNCTION AddToSet(NextValues MEDIUMTEXT, CurrentSet MEDIUMTEXT) RETURNS MEDIUMTEXT DETERMINISTIC BEGIN IF LENGTH(CurrentSet) = 0 OR CurrentSet is NULL THEN RETURN NextValues; ELSEIF LENGTH(NextValues) = 0 OR NextValues is NULL THEN RETURN CurrentSet; # AddToSet should be symmetrical ELSE RETURN CONCAT(CurrentSet,',',NextValues); END IF; END ; nested exception is com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 1
        at org.modelcatalogue.core.InitCatalogueService.setupStoredProcedures(InitCatalogueService.groovy:201)
        at BootStrap$_closure1.doCall(BootStrap.groovy:67)
        at grails.util.Environment.evaluateEnvironmentSpecificBlock(Environment.java:327)
        at grails.util.Environment.executeForEnvironment(Environment.java:320)
        at grails.util.Environment.executeForCurrentEnvironment(Environment.java:296)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 1
        at com.mysql.jdbc.Util.handleNewInstance(Util.java:411)
        at com.mysql.jdbc.Util.getInstance(Util.java:386)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:1054)
        at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:4120)
        at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:4052)
        at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2503)
        at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2664)
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2809)
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2758)
        at com.mysql.jdbc.StatementImpl.execute(StatementImpl.java:894)
        at com.mysql.jdbc.StatementImpl.execute(StatementImpl.java:732)
        ... 9 more
```

**Comment by passionfruit18** [comment]
For the error, is ELSEIF an ok syntax? Could it be that different versions of SQL don't like this syntax?

**Comment by musketyr** [comment]
looks good to me. does running the procedure from within IntelliJ or MySQL Workbench also give you any error?

**Comment by passionfruit18** [comment]
no, ELSEIF is fine in MySQL Workbench.

**Comment by passionfruit18** [comment]
ok so it was to do with the comments, the loading command wanted comments starting with -- rather than #.

File: ModelCatalogueCorePluginTestApp/src/groovy/org/modelcatalogue/core/mysql-procedures.sql (hunk 1 above)
Comment (passionfruit18) [inline ModelCatalogueCorePluginTestApp/src/groovy/org/modelcatalogue/core/mysql-procedures.sql]: ok I've put them above.

**Pull request merged**